### ISSUE TITLE
Compiling fails

### ISSUE THREAD

**Issue by roth-a** [modmido/psiquasp#1]
I followed the install instructions precisely, but get the following error when compiling PsiQuaSP with
`make`.

The error is: 
```
petsc//include/petscmath.h:320:9: error: ‘PetscComplex’ does not name a type
 typedef PetscComplex PetscScalar;
```
from this follow more errors like
```
 error: ‘PetscIsInfScalar’ declared as an ‘inline’ variable
 PETSC_STATIC_INLINE PetscBool PetscIsInfScalar(PetscScalar v) {return PetscIsInfReal(PetscAbsScalar(v));}

```

**Comment by modmido** [comment]
Hello Alexander,

to me it seems that there has to be something wrong with your petsc 
build, since it appears that the PetscComplex type was not declared, 
which it should. Have you set the --with-scalar-type=complex flag in 
your petsc configure stage?
Sometimes when there is a typo in the configure command petsc may not 
use the flag and not tell you. You can check in the configure.log file 
in your petsc directory, there in line 13 petsc lists all the flags that 
are actually used for the configure stage. If that flag is not present 
there then you should rerun the petsc build.

If that does not fix your problem then please provide me with more 
information, like your operating system, the petsc version you are using 
and the full output of the error message and your petsc configure log, 
since otherwise I cannot help you.

Cheers,

Michael

PS: thanks for downloading the code, I hope you like it.


Am 04.09.2017 um 12:58 schrieb Alexander:

**Comment by roth-a** [comment]
Thanks for the quick answer. configure.log shows complex is included, but it also shows lots of errors below, like
`/usr/bin/ld: cannot find -lpetsc-ufod4vtr9mqHvKIQiVAm`  I will first go through the errors myself and check if I can figure out what the problem is.

**Comment by roth-a** [comment]
I attached the configure log file [configure.txt](https://github.com/modmido/psiquasp/files/1276759/configure.txt) and the config [output.txt](https://github.com/modmido/psiquasp/files/1276793/output.txt)
Petsc make gives [make.txt](https://github.com/modmido/psiquasp/files/1276767/make.txt)
again when trying to "make" psiquasp, I get
```
error: ‘PetscComplex’ does not name a type
 typedef PetscComplex PetscScalar;

```

I'm a little lost. One thing I noticed in the Petsc config output was the empty field
```
cmake:
  Arch:     
```
Is it supposed to be like that?

**Comment by modmido** [comment]
Hi there,
the empty arch line is ok, but in your make.txt file make says that there is nothing to do in your petsc folder, so basically the library is never created, which explains why PetscComplex is not defined, since nothing gets defined. At this point in the make log there should be a long output about the countless *.o files created by the compiler. To be honest I never encountered this problem before. What is your 

`make test`

output? Do you have previously build petsc and did not delete the PETSC_ARCH folder before rebuilding? This is recommended.

**Comment by roth-a** [comment]
Yes I have built it before, explaining why there was nothing to do. I will try again after deleting the PETSC_ARCH folder.

**Comment by roth-a** [comment]
after configuring and making PETSC I have run  `make test` with the output:
```                                                                                 
Running test examples to verify correct installation
Using PETSC_DIR=/home/test/code-div/petsc and PETSC_ARCH=arch-linux2-c-debug
C/C++ example src/snes/examples/tutorials/ex19 run successfully with 1 MPI process
C/C++ example src/snes/examples/tutorials/ex19 run successfully with 2 MPI processes
Fortran example src/snes/examples/tutorials/ex5f run successfully with 1 MPI process
Completed test examples
=========================================
Now to evaluate the computer systems you plan use - do:
make PETSC_DIR=/home/test/code-div/petsc PETSC_ARCH=arch-linux2-c-debug streams

```
And also `make PETSC_DIR=/home/test/code-div/petsc PETSC_ARCH=arch-linux2-c-debug streams` works.

However building psiquasp still fails with the same error.

I know it's difficult to analyze such an error. Could you try to follow your installation steps on a virtual ubuntu machine?

**Comment by roth-a** [comment]
[make.txt](https://github.com/modmido/psiquasp/files/1277359/make.txt)
[options.txt](https://github.com/modmido/psiquasp/files/1277358/options.txt)

**Comment by martenrichter** [comment]
May be add an include of 
`#include <quadmath.h>`
before the petsc include in the file, where it fails.

**Comment by roth-a** [comment]
Where is quadmath.h located and, in which file should i include ` #include <quadmath.h>` ?

**Comment by martenrichter** [comment]
I do not know where quadmath.h is located on your system. I only know that the petsc build includes libquadmath for handling the 128 bit floating point types. The building error, that you get tells that the type for 128 bit complex floating point type is not included, which is in your case defined through quadmath. 
That is why suggest to include quadmath.h in the file where the compilation fails before the inclusion of the petsc header files. 
Another possibility is, that the include directory for quadmath is not found during the build of psyquasp.
On most linux systems, you should be able to locate quadmath.h using "locate quadmath.h"
On our debian jessie system it is under "/usr/lib/gcc/x86_64-linux-gnu/4.8/include/quadmath.h" or "/usr/lib/gcc/x86_64-linux-gnu/4.9/include/quadmath.h"

**Comment by roth-a** [comment]
Yes, locate quadmath.h gives:
```
/usr/lib/gcc/x86_64-linux-gnu/4.9/include/quadmath.h
/usr/lib/gcc/x86_64-linux-gnu/5/include/quadmath.h
/usr/lib/gcc/x86_64-linux-gnu/6/include/quadmath.h
```
I included `#include <quadmath.h>`  in all *.hpp files in psiquasp, before all other `include` statements (I have no idea in which file  the error occurs). The make error stays the same. I'd like to to more systematic debugging, but 'm unexperienced in c++ programming.
Maybe a quick test in a fresh linux/ubuntu using virtualbox, would reproduce the error on your end.

**Comment by martenrichter** [comment]
`Maybe a quick test in a fresh linux/ubuntu using virtualbox, would reproduce the error on your end.`
We can not do this, since we do not have root access to install software on our machines and it is highly unlikely, that we match your package configuration. 
Can you please post the make output now with all errors and warnings....

**Comment by roth-a** [comment]
Here is the output of the make command (using a freshly cloned psiquasp):
[make.output.txt](https://github.com/modmido/psiquasp/files/1280579/make.output.txt)

**Comment by martenrichter** [comment]
Ok, I think it is clear, that PetscComplex is not properly defined. 
In order to figure out, why no complex numbers are defined please insert the line
`#warning "Branch entered at line"`
in petscmath.h after lines 180, 196, 212, 223, 243, 263 and then post again the make file.
Then I will know, which branch the preprocessor takes, and then I may have another idea....

**Comment by roth-a** [comment]
I included ` #warning "Branch entered at line"` in the mentioned lines. I deleted the arch-linux2-c-debug folder in petsc, did
```
python2 './configure' '--with-cc=gcc' '--with-cxx=g++' '--with-fc=gfortran' '--download-mpich' '--with-clanguage=c' '--with-scalar-type=complex' '--with-precision=__float128' '--download-f2cblaslapack'
make PETSC_DIR=/home/test/code-div/petsc PETSC_ARCH=arch-linux2-c-debug all
``` 
Compiling it gives endless statements like
```
 #warning "Branch entered at line"
  ^~~~~~~
          CC arch-linux2-c-debug/obj/src/ksp/pc/impls/pbjacobi/pbjacobi.o
In file included from /home/test/code-div/petsc/include/petscsys.h:439:0,
                 from /home/test/code-div/petsc/include/petscis.h:7,
                 from /home/test/code-div/petsc/include/petscvec.h:9,
                 from /home/test/code-div/petsc/include/petscmat.h:6,
                 from /home/test/code-div/petsc/include/petscpc.h:6,
                 from /home/test/code-div/petsc/include/petscksp.h:6,
                 from /home/test/code-div/petsc/include/petsc/private/pcimpl.h:5,
                 from /home/test/code-div/petsc/src/ksp/pc/impls/pbjacobi/pbjacobi.c:7:
/home/test/code-div/petsc/include/petscmath.h:269:2: warning: #warning "Branch entered at line" [-Wcpp]
```
Make test gives [maketest.output.txt](https://github.com/modmido/psiquasp/files/1280803/maketest.output.txt). 
After that I ran make in psiquasp. The output is [make.output.txt](https://github.com/modmido/psiquasp/files/1280801/make.output.txt)

**Comment by martenrichter** [comment]
Prefect, now, we know, what happens.
> The output is make.output.txt
> #warning "Branch entered at line" does not appear at all in making of psiquasp, which puzzles me.
This means, that the macros set during for the petsc architecture are not set for the psiquasp build.
That means, that the file in your architecture include dir petscconf.h is not properly included.

Can you please place an
`#error "I am included"`
in your archtitectures include dir petscconf.h.
Then you should rebuild psiquasp, which should fail with "I am included", if it does not fail, there are two options: "petscconf.h" is not included at all, or you have a second petscconf.h some where in the include path, which is interfering, so search for this file name. You can identify the file, which is included by placing different error messages.

**Comment by roth-a** [comment]
Making psiquasp gives now
```
petscconf.h:1:2: error: #error "I am included"
 #error "I am included"
```

**Comment by martenrichter** [comment]
Ok, so it is included? Then please post the file...

**Comment by roth-a** [comment]
[petscconf.h.txt](https://github.com/modmido/psiquasp/files/1281018/petscconf.h.txt)
(I removed ` #error "I am included"`)

**Comment by martenrichter** [comment]
Ok, looks fine as well.
Please add "-dD" to the CXXFLAGS in your options.mk . And then build psiquap again and post the output. It should show the result of the preprocessor and give additional clues.

**Comment by roth-a** [comment]
Ok, done: [make.txt](https://github.com/modmido/psiquasp/files/1281099/make.txt)

**Comment by martenrichter** [comment]
Sorry, the option did not have any effect. Please use the option "-save-temps" instead. Now a file index.ii should be generated, please post this file.

**Comment by roth-a** [comment]
[index.ii.txt](https://github.com/modmido/psiquasp/files/1281361/index.ii.txt)

**Comment by martenrichter** [comment]
We have reproduced the problem and are now working on a fix.

**Comment by modmido** [comment]
So the problem is the lacking support for complex float128 in combination with c++ in petsc. We used a patched version of petsc, that enforces the use of c style complex numbers for that case, and forgot about that. So the easiest way to circumvent this problem is to build petsc and psiquasp with double precision instead of float128. In our case this works fine with standard petsc. This might lead to reduced stability but it should be fine in almost all cases, only if systems become large and really long time dynamics are computed this may cause problems. We are in contact with the petsc developers to include this functionality, until then we will upload a git patch for the extended precision support. Sorry for the inconvenience.

**Comment by roth-a** [comment]
ok. thanks.

**Comment by roth-a** [comment]
I followed the instructions for version 3.7.6
```
Note: checking out 'v3.7.6'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD ist jetzt bei 7587a4b... Increase patchlevel to 3.7.6

```
 with the patch. 
`git apply petscpatch.patch
`  (no error message)
However then I cannot configure petsc anymore:
`python2 './configure' '--with-cc=gcc' '--with-cxx=g++' '--with-fc=gfortran' '--download-mpich' '--with-clanguage=c' '--with-scalar-type=complex' '--with-precision=__float128' '--download-f2cblaslapack'
`
```

===============================================================================
             Configuring PETSc to compile on your system                       
===============================================================================
TESTING: checkCCompiler from config.setCompilers(config/BuildSystem/config/setCompilers.py:553)                                                                                                                                                    
===============================================================================                                                                                                                                                                          Trying to download http://www.mpich.org/static/downloads/3.1.3/mpich-3.1.3.tar.gz for MPICH                                                                                                                                                  ===============================================================================                                                                                                                                                                    ===============================================================================                                                                                                                                                                          Running configure on MPICH; this may take several minutes                                                                                                                                                                                    ===============================================================================                                                                                                                                                                    ===============================================================================                                                                                                                                                                          Running make on MPICH; this may take several minutes                                                                                                                                                                                         ===============================================================================                                                                                                                                                                                                                                                                                                                                                                                                                       
===============================================================================                                                                                                                                                                          Running make install on MPICH; this may take several minutes                                                                                                                                                                                 ===============================================================================                                                                                                                                                                    ===============================================================================                                                                                                                                                                          Trying to download git://https://bitbucket.org/petsc/pkg-sowing.git for SOWING                                                                                                                                                               ===============================================================================                                                                                                                                                                    ===============================================================================                                                                                                                                                                          Running configure on SOWING; this may take several minutes                                                                                                                                                                                   ===============================================================================                                                                                                                                                                    ===============================================================================                                                                                                                                                                          Running make on SOWING; this may take several minutes                                                                                                                                                                                        ===============================================================================                                                                                                                                                                    ===============================================================================                                                                                                                                                                          Running make install on SOWING; this may take several minutes                                                                                                                                                                                ===============================================================================                                                                                                                                                                    TESTING: checkSharedLibrary from config.packages.sowing(config/BuildSystem/config/package.py:738)                                                                                                                                                  Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/mat/impls/aij/seq/superlu/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/mat/impls/aij/seq/seqcusparse/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/mat/impls/aij/seq/seqcusp/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/mat/impls/aij/seq/mkl_pardiso/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/mat/impls/aij/mpi/mpicusparse/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/mat/impls/aij/mpi/mumps/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/mat/impls/aij/mpi/mkl_cpardiso/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/mat/impls/aij/mpi/mpicusp/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/vec/vec/impls/seq/seqcuda/makefile
Error! missing file: /home/alexander/work/Promotion/code-div/petsc/src/vec/vec/impls/seq/seqcusp/makefile
===============================================================================                                                                                                                                                                          Trying to download http://ftp.mcs.anl.gov/pub/petsc/externalpackages/f2cblaslapack-3.4.2.q1.tar.gz for F2CBLASLAPACK                                                                                                                         ===============================================================================                                                                                                                                                                    ===============================================================================                                                                                                                                                                          Compiling F2CBLASLAPACK; this may take several minutes                                                                                                                                                                                       ===============================================================================                                                                                                                                                                    ===============================================================================                                                                                                                                                                          Installing F2CBLASLAPACK                                                                                                                                                                                                                     ===============================================================================                                                                                                                                                                    TESTING: configureViewFromOptions from PETSc.Configure(config/PETSc/Configure.py:1025)                                                                                                                                                             *******************************************************************************
         UNABLE to CONFIGURE with GIVEN OPTIONS    (see configure.log for details):
-------------------------------------------------------------------------------
PETSc makefiles contain mistakes or files are missing on filesystem.
Makefile contains file not on filesystem: src/sys/classes/draw/impls/x: ['xioerr.c']
Possible reasons:
        1. Files were deleted locally, try "hg revert filename" or "git checkout filename".
        2. Files were deleted from repository, but were not removed from makefile. Send mail to petsc-maint@mcs.anl.gov.
        3. Someone forgot to "add" new files to the repository. Send mail to petsc-maint@mcs.anl.gov.
*******************************************************************************

```

**Comment by modmido** [comment]
Hi, this is a strange error message, since the patch does not alter any of the missing files or even the directories in which these files are. In my petsc version (patched and unpatched v3.7.6) these files are present. Can you undo the changes (return to v3.7.6), i.e. by
`git checkout -b cxxcomplexpatch`
`git commit -a`
`git checkout v3.7.6`
and then try to install petsc then (e.g. with double precision) and see whether the error remains the same? If yes, there is some problem with your petsc version. Then try deleting the whole folder and reclone it from bitbucket. Thank you.

**Comment by roth-a** [comment]
OK, I'll try it later. First I try to try the non-patch install instructions.

**Comment by modmido** [comment]
ok, glad that the normal installation works now. I tried to reproduce your error messages using two different petsc clones but in my case everything works fine. Did you get the same error messages with the freshly cloned petsc? Are these files actually missing? They should not be missing and are certainly not removed by our patch. Could you send us your configure.log of the failed installation? In any case the petsc developers said, that the c++ complex float128 support will be included in their 3.8 release, which will probably happen in the next couple of months.

**Comment by roth-a** [comment]
I followed again the patch-instructions from a freshly cloned repo (in between I copied the patch file):
```
git clone https://bitbucket.org/petsc/petsc.git
cd petsc
git checkout v3.7.6
git apply petscpatch.patch
python2 './configure' '--with-cc=gcc' '--with-cxx=g++' '--with-fc=gfortran' '--download-mpich' '--with-clanguage=c' '--with-scalar-type=complex' '--with-precision=__float128' '--download-f2cblaslapack'
```
This now configures and compiles correctly. (Sorry for the confusion, I have no idea why i had the error before)
And psiquasp also compiles.
Thanks for all the help!
